### ISSUE TITLE
[Feature request] Page binding (left/right edge)

### ISSUE THREAD

**Issue by oatmealm** [pikepdf/pikepdf#111]
I'm wondering it's possible to support left/right edge binding in pikepdf? It should allow correct flow of PDFs which are predominately written in right-tol-left languages, such as Arabic, Farsi, Hebrew etc. I've seen it implemented in apps based of PSPDFkit, which are not free software.

**Comment by dreua** [comment]
I don't understand what you mean by binding. Is it about the order in which pages are displayed, i.e. first page would be top right instead of top left?

**Comment by oatmealm** [comment]
> I don't understand what you mean by binding.

Yes. I think it's called binding. Anyways, yes. It's ordering of the page in fact. It's meaningful when scrolling  horizontally.

**Comment by dreua** [comment]
Direction of the iconview can be easily set:
`self.iconview.set_direction(Gtk.TextDirection.RTL)`
https://github.com/dreua/pdfarranger/blob/905d47df81a9cbc64dd50674a8e4710aff5fe3a2/pdfarranger/pdfarranger.py#L441

This aligns all the pages to the top right and the pages are displayed from right to left (page 1 top right, 2 on the left of it and so on).

1. I wonder if this breaks any of our other logic. (Needs testing.)
2. We need a way to automatically set/unset this. Do I understand that you do not want an option for this but it should be read from the PDF itself? Could you provide such a PDF for testing? 
3. Further questions: What should happen if some but not all of the added pages are Right to Left? Do we need options to overwrite the display or set the direction of the PDFs?

**Comment by oatmealm** [comment]
Thanks for looking into it.

Is this a GUI option currently in PDFArranger's UI, or do you mean it could be set progrmmatically?


I don't remember seeing such an option anywhere other than PDF Viewer (the name of the default implementation based of PSPDFkit on iOS). 

CORRECTION: I just opened a PDF in Evince (Gnome) and I see there's an option to set a document as Right to Left.... never seen this before, but it doesn't seem to work. When I change rtl to ltr or ltr to rtl and save the document and then view it in PDFViewer of the iPad (PSPDFkit) it doesn't seem to have any effect. It's probably just for the display and doesn't chnage the document itself.

As for mixed script, I guess the default ltr is perfectly acceptable, unless this "binding" property was already set by the user? I don't think it's trivial/possible to guess it progrmaticaly? I know that some paper I get from my university come already set for rtl binding, but it's rare.

This is an example of a PDF I've created (print to PDF) and set in PSPDFkit to bind as rtl.
[בין_מקומי_לעולמי,_פנתר.pdf](https://github.com/pikepdf/pikepdf/files/4945869/_._._.pdf)
![2020-07-20 08-25-19-1](https://user-images.githubusercontent.com/68159077/87906586-3354f400-ca63-11ea-8264-e4a5f7f42fdf.png)

**Comment by jbarlow83** [comment]
This can be done by capturing a page as a Form XObject (PDF parlance for a group of objects)... rough sketch of solution:

```python

pdf = pikepdf.open('...')
dictpage0 = pdf.pages[0]
page0 = pikepdf.Page(dictpage0)
frmx = page0.as_form_xobject()

frmx2 = ... # extract the second page as form xobject, like above

resized = ... # calculate size of new mediabox here

```

and then constructing a new page of double width that plots the new two form Form XObjects

```python
# Divide page in half vertically, draw the /PageLeft object in the first half and /PageRight in the second
newpage_contents = pikepdf.Stream(pdf, b"""
q
0.5 0 0 1 0 0 cm
/PageLeft Do
0.5 0 0 1 0.5 0 cm
/PageRight Do
Q
""")

newpage = pikepdf.Dictionary(
    Type=pikepdf.Name.Page, 
    MediaBox=resized, 
    Resources=pikepdf.Dictionary(
        XObject=pikepdf.Dictionary(
            PageLeft=frmx, PageRight=frmx2
        )
    )
    Contents=newpage_contents
)

pdf.pages.append(newpage)
```

This is a rough example without some details worked out, like whether to where to put the new page and delete the old page, and dealing with the precedence of Cropbox vs Mediabox.

**Comment by oatmealm** [comment]
Just to make sure, is there such a property (i.e. binding=rtl) in the PDF spec?

**Comment by dreua** [comment]
Haha omg I mixed up repositories, thought this was pdfarranger, sorry, will hide my off topic comments :see_no_evil:

**Comment by dreua** [comment]
> Is this a GUI option currently in PDFArranger's UI, or do you mean it could be set progrmmatically?

No we don't have RTL support in pdfarranger. I was on the false assumption that this was a issue raised against pdfarranger and made this just as a quick proof of concept to show that the Gtk component we use to display pages supports rtl quite easily. We would need the logic of turning it on/off and/or reading/writing it from/to PDFs. If you actually want rtl support in pdfarranger, please raise an issue there.

**Comment by jbarlow83** [comment]
You can set `/Catalog /ViewerPreferences /Direction` to `/R2L`; see section 12.2 in the PDF reference manual. For pikepdf: 
```python
if not pdf.root.ViewerPreferences:
    pdf.root.ViewerPreferences = pikepdf.Dictionary()
pdf.root.ViewerPreferences.Direction = pikepdf.Name.R2L
```

As the manual says this has no effect on the PDF other than letting the viewer know that the file should be treated as RTL page order. It's up to the PDF generator to lay out the PDF contents properly for RTL.

**Comment by dreua** [comment]
@jbarlow83 I think your first sketch (with XObjects) would take two pages [1] [2] and combine them into one lager page with the first page on the right and the second on the left [2 1] correct? I'm not sure this is what @oatmealm wants.
As I understand it, this is more about how pages are displayed and therefore partly out of scope here in pikepdf (since it doesn't display anything). The role of pikepdf here could be merely to provide easy access to read/write the direction from/to PDFs.

Your latest comment just appeared: Yes, I think that is what we need here, so how I understand it this feature is already there. Thanks for your expertise!
